cpp/src/arrow/flight/sql/odbc/odbc_api.cc

```diff
@@ -42,7 +43,26 @@ namespace arrow
       }
 
       case SQL_HANDLE_DBC: {
-        return SQL_INVALID_HANDLE;
+        using ODBC::ODBCConnection;
+        using ODBC::ODBCEnvironment;
+
+        *result = SQL_NULL_HDBC;
+
+        ODBCEnvironment* environment = reinterpret_cast<ODBCEnvironment*>(parent);
+
+        if (!environment) {
+          return SQL_INVALID_HANDLE;
+        }
+
+        std::shared_ptr<ODBCConnection> conn = environment->CreateConnection();
+
+        if (!conn) {
+          return environment->GetDiagnostics().GetNativeError(0);
+        }
+
+        *result = reinterpret_cast<SQLHANDLE>(&conn);
```

Comment:
The conn shared_ptr goes out of scope here, which will cause the wrapped OdbcConnection object to get deleted (since it is the sole owner of it). Returning a pointer to this current shared_ptr won't extend its lifetime since it is allocated on the stack. You have a few options here:
- Return a clone of the shared_ptr allocated on the heap using new and store it in *result.
- Have something else with a longer lifetime (such as either the driver or ODBCEnvironment) hold a clone of the shared_ptr.
- Use raw pointers instead of shared_ptrs. Detach the raw ODBCConnection pointer from the shared_ptr and return it instead.

Note that if you return a shared_ptr to the caller then they pass it back, you can't cast it to an ODBCConnection* and delete it. You have to cast it to a shared_ptr*.

Reply:
OK sure. But then we should remove the extra shared_ptr held in ODBCEnvironment when freeing the connection handle so the list doesn't keep growing in size (or does releaseConnection() do this?)